HDF5 issues in CMAKE

Hi again @sigurdstorve 

Thanks for the detailed explanation of the software on the other thread :+1: 

I am trying to cmake the file now. I actually followed the following steps 
$ cmake .
$ sudo make install

But, later I tried to run the python script $python linear_scan_model.py - [along with the arguments mentioned in the HOW_TOPYTHON.md]

As expected it failed because, I didn't make the file like you suggested to do, i.e go into the cmake GUI and switch some variables to ON and subsequently when I tried to generate - It gave me the following error. 
` CMake Error at CMakeLists.txt:42 (message):
   Imported target hdf5-shared is not available
`  
My configuration file looks like follows:
`
```
 BCSIM_BUILD_BENCHMARK_CODE       OFF 
                                                                            
 BCSIM_BUILD_EXAMPLES             OFF
                                                                             
 BCSIM_BUILD_PYTHON_INTERFACE     ON                                                                              

 BCSIM_BUILD_QT5_GUI              OFF                                                                             

 BCSIM_BUILD_UNITTEST             OFF                                                                             

 BCSIM_BUILD_UTILS                ON                                                                              

 BCSIM_ENABLE_CUDA                OFF                                                                             

 BCSIM_ENABLE_NAN_CHECK           ON                                                                              

 BCSIM_ENABLE_OPENMP              ON                                                                              

 BUILD_SHARED_LIBS                ON                                                                              

 CMAKE_BUILD_TYPE                                                                                                 

 CMAKE_INSTALL_PREFIX             /usr/local                                                                      

 HDF5_C_LIBRARY_dl                /usr/lib/x86_64-linux-gnu/libdl.so                                              

 HDF5_C_LIBRARY_hdf5              /usr/lib/x86_64-linux-gnu/libhdf5.so                                            

 HDF5_C_LIBRARY_m                 /usr/lib/x86_64-linux-gnu/libm.so                                               

 HDF5_C_LIBRARY_pthread           /usr/lib/x86_64-linux-gnu/libpthread.so                                         

 HDF5_C_LIBRARY_z                 /usr/lib/x86_64-linux-gnu/libz.so                                               

 HDF5_DIR                         HDF5_DIR-NOTFOUND                                                               

```
`
As per hdf5, I generally use hdf5 files with Matlab (in Ubuntu 16.04) and also with other programming languages and it all works fine.
I could not understand what is raising this issue.
Could you please help me out with this?

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -34,15 +34,12 @@ include_directories(${PROJECT_SOURCE_DIR})
 find_package(Boost REQUIRED)
 
 if (BCSIM_BUILD_UTILS OR BCSIM_BUILD_QT5_GUI)
-    # Note: Using the new-style find_package() from HDF5 which
-    # makes available imported targets.
-    # Then it is no longer needed to do manually do include_directories()
+    # Note: Prefer the new-style find_package() for HDF5 which creates imported targets.
+    # Then it is no longer needed to manually do include_directories().
     find_package(HDF5 REQUIRED COMPONENTS C CXX)
-    if (NOT TARGET hdf5-shared)
-        message(FATAL_ERROR "Imported target hdf5-shared is not available")
-    endif()
-    if (NOT TARGET hdf5_cpp-shared)
-        message(FATAL_ERROR "Imported target hdf5_cpp-shared is not available")
+    if (NOT TARGET hdf5-shared OR NOT TARGET hdf5_cpp-shared)
+        message(WARNING "Could not find imported HDF5 targets. Using module-mode instead.")
+        include_directories(${HDF5_INCLUDE_DIRS}) 
     endif()
 endif()
 

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -37,7 +37,7 @@ if (BCSIM_BUILD_UTILS OR BCSIM_BUILD_QT5_GUI)
     # Note: Using the new-style find_package() from HDF5 which
     # makes available imported targets.
     # Then it is no longer needed to do manually do include_directories()
-    find_package(HDF5 REQUIRED)
+    find_package(HDF5 REQUIRED COMPONENTS C CXX)
     if (NOT TARGET hdf5-shared)
         message(FATAL_ERROR "Imported target hdf5-shared is not available")
     endif()

--- a/src/python/CMakeLists.txt
+++ b/src/python/CMakeLists.txt
@@ -26,8 +26,8 @@ target_link_libraries(pyrfsim
                       Boost::python
                       Boost::system
                       LibBCSim
-                      hdf5-static
-                      hdf5_cpp-static
+                      hdf5-shared
+                      hdf5_cpp-shared
                       )
 set_target_properties(pyrfsim PROPERTIES PREFIX "")
 if (WIN32)

--- a/src/python/CMakeLists.txt
+++ b/src/python/CMakeLists.txt
@@ -26,9 +26,13 @@ target_link_libraries(pyrfsim
                       Boost::python
                       Boost::system
                       LibBCSim
-                      hdf5-shared
-                      hdf5_cpp-shared
                       )
+if (TARGET hdf5-shared AND TARGET hdf5_cpp-shared)
+    target_link_libraries(pyrfsim hdf5-shared hdf5_cpp-shared)
+else()
+    target_link_libraries(pyrfsim ${HDF5_LIBRARIES})
+endif()
+                      
 set_target_properties(pyrfsim PROPERTIES PREFIX "")
 if (WIN32)
 set_target_properties(pyrfsim PROPERTIES SUFFIX ".pyd")

--- a/src/qt5gui/CMakeLists.txt
+++ b/src/qt5gui/CMakeLists.txt
@@ -92,10 +92,15 @@ target_link_libraries(BCSimGUI
                       Boost::boost
                       LibBCSim
                       LibBCSimUtils # BCSimCUDA also?
-                      hdf5-shared
-                      hdf5_cpp-shared
                       trianglemesh3d
                       )
+
+if (TARGET hdf5-shared AND TARGET hdf5_cpp-shared)
+    target_link_libraries(BCSimGUI hdf5-shared hdf5_cpp-shared)
+else()
+    target_link_libraries(BCSimGUI ${HDF5_LIBRARIES})
+endif()
+
 if (WIN32)
     target_link_libraries(BCSimGUI Qt5::WinMain)
 endif()

--- a/src/qt5gui/CMakeLists.txt
+++ b/src/qt5gui/CMakeLists.txt
@@ -92,8 +92,8 @@ target_link_libraries(BCSimGUI
                       Boost::boost
                       LibBCSim
                       LibBCSimUtils # BCSimCUDA also?
-                      hdf5-static
-                      hdf5_cpp-static
+                      hdf5-shared
+                      hdf5_cpp-shared
                       trianglemesh3d
                       )
 if (WIN32)

--- a/src/utils/CMakeLists.txt
+++ b/src/utils/CMakeLists.txt
@@ -25,8 +25,8 @@ list(APPEND UTILS_LIBRARY_SOURCE_FILES
 
 add_library(LibBCSimUtils ${UTILS_LIBRARY_SOURCE_FILES})
 target_link_libraries(LibBCSimUtils
-                      hdf5-static
-                      hdf5_cpp-static
+                      hdf5-shared
+                      hdf5_cpp-shared
                       Boost::boost
                       )
 

--- a/src/utils/CMakeLists.txt
+++ b/src/utils/CMakeLists.txt
@@ -25,11 +25,15 @@ list(APPEND UTILS_LIBRARY_SOURCE_FILES
 
 add_library(LibBCSimUtils ${UTILS_LIBRARY_SOURCE_FILES})
 target_link_libraries(LibBCSimUtils
-                      hdf5-shared
-                      hdf5_cpp-shared
                       Boost::boost
                       )
 
+if (TARGET hdf5-shared AND TARGET hdf5_cpp-shared)
+    target_link_libraries(LibBCSimUtils hdf5-shared hdf5_cpp-shared)
+else()
+    target_link_libraries(LibBCSimUtils ${HDF5_LIBRARIES})
+endif()
+                      
 if (BCSIM_ENABLE_CUDA)
     # Needed because we're going to interact with the CUDA runtime API
     # to query the number and type of available GPUs.

